File: Reqnroll/Shared/WIMP.App/Data/Db/DataRepository.cs
/*
 * NOTE: This application ("WIMP - Where Is My Pizza") is provided solely to
 * demonstrate the Behavior-Driven Development scenario automation patterns
 * described in the book "Effective Behavior-Driven Development" by
 * Gaspar Nagy and Seb Rose.
 *
 * It is NOT a complete or production-ready implementation. It deliberately
 * uses shortcuts and simplifications (e.g. authentication, data storage,
 * error handling, security) that are NOT suitable for a real application.
 * Do not use this code as a basis for production software.
 */

/*
 * NOTE: This application ("WIMP - Where Is My Pizza") is provided solely to
 * demonstrate the Behavior-Driven Development scenario automation patterns
 * described in the book "Effective Behavior-Driven Development" by
 * Gaspar Nagy and Seb Rose.
 *
 * It is NOT a complete or production-ready implementation. It deliberately
 * uses shortcuts and simplifications (e.g. authentication, data storage,
 * error handling, security) that are NOT suitable for a real application.
 * Do not use this code as a basis for production software.
 */

using System.Text.Json;

using Microsoft.EntityFrameworkCore;

using WIMP.App.Data.Db.DbRecords;
using WIMP.App.Models;

namespace WIMP.App.Data.Db;

public class DataRepository(IDbContextFactory<WimpDbContext> dbContextFactory) : IDataRepository
{
    public Customer? GetCustomerByName(string customerName)
    {
        using var db = dbContextFactory.CreateDbContext();
        var customer = db.CustomerRows
            .FirstOrDefault(c => c.Name.ToUpper() == customerName.ToUpper());

        return customer is null
            ? null
            : new Customer
            {
                Name = customer.Name,
                Email = customer.Email
            };
    }

    public void InsertCustomer(Customer customer)
    {
        using var db = dbContextFactory.CreateDbContext();
        db.CustomerRows.Add(new CustomerDbRecord
        {
            Id = Guid.NewGuid(),
            Name = customer.Name,
            Email = customer.Email
        });
        db.SaveChanges();
    }

    public void InsertSession(Session session)
    {
        using var db = dbContextFactory.CreateDbContext();
        db.SessionRows.Add(new SessionDbRecord
        {
            Token = session.Token,
            CustomerName = session.CustomerName
        });
        db.SaveChanges();
    }

    public void DeleteSession(string token)
    {
        using var db = dbContextFactory.CreateDbContext();
        var session = db.SessionRows.FirstOrDefault(s => s.Token == token);
        if (session is not null)
        {
            db.SessionRows.Remove(session);
            db.SaveChanges();
        }
    }

    public Session? GetSessionByToken(string token)
    {
        using var db = dbContextFactory.CreateDbContext();
        var session = db.SessionRows.FirstOrDefault(s => s.Token == token);
        return session is null ? null : new Session(session.Token, session.CustomerName);
    }

    public void InsertCoupon(Coupon coupon)
    {
        using var db = dbContextFactory.CreateDbContext();
        db.CouponRows.Add(new CouponDbRecord
        {
            Id = Guid.NewGuid(),
            Code = coupon.Code,
            CustomerEmail = coupon.CustomerEmail,
            SentAt = coupon.SentAt
        });
        db.SaveChanges();
    }

    public IReadOnlyCollection<Coupon> GetCouponsByEmail(string customerEmail)
    {
        using var db = dbContextFactory.CreateDbContext();
        return db.CouponRows
            .Where(c => c.CustomerEmail.ToUpper() == customerEmail.ToUpper())
            .Select(c => new Coupon
            {
                Code = c.Code,
                CustomerEmail = c.CustomerEmail,
                SentAt = c.SentAt
            })
            .ToArray();
    }

    public void InsertNotification(Notification notification)
    {
        using var db = dbContextFactory.CreateDbContext();
        db.NotificationRows.Add(new NotificationDbRecord
        {
            Id = Guid.NewGuid(),
            CustomerName = notification.CustomerName,
            Message = notification.Message,
            SentAt = notification.SentAt
        });
        db.SaveChanges();
    }

    public IReadOnlyCollection<Notification> GetNotificationsByCustomerName(string customerName)
    {
        using var db = dbContextFactory.CreateDbContext();
        return db.NotificationRows
            .Where(c => c.CustomerName.ToUpper() == customerName.ToUpper())
            .Select(c => new Notification
            {
                CustomerName = c.CustomerName,
                Message = c.Message,
                SentAt = c.SentAt
            })
            .ToArray();
    }

    public void InsertPromotion(Promotion promotion)
    {
        using var db = dbContextFactory.CreateDbContext();
        db.PromotionRows.Add(new PromotionDbRecord
        {
            Id = Guid.NewGuid(),
            Name = promotion.Name,
            IsActive = promotion.IsActive
        });
        db.SaveChanges();
    }

    public Promotion? GetPromotionByName(string name)
    {
        using var db = dbContextFactory.CreateDbContext();
        var promotion = db.PromotionRows
            .FirstOrDefault(c => c.Name.ToUpper() == name.ToUpper());

        return promotion is null
            ? null
            : new Promotion
            {
                Name = promotion.Name,
                IsActive = promotion.IsActive
            };
    }

    public Promotion UpdatePromotion(Promotion promotion, bool? isActive = null)
    {
        if (isActive is not null)
        {
            promotion.IsActive = isActive.Value;
        }

        using var db = dbContextFactory.CreateDbContext();
        var dbPromotion = db.PromotionRows
            .FirstOrDefault(c => c.Name.ToUpper() == promotion.Name.ToUpper());

        if (dbPromotion is not null)
        {
            dbPromotion.IsActive = promotion.IsActive;
            db.SaveChanges();
        }

        return promotion;
    }

    public Order InsertOrder(Order order)
    {
        using var db = dbContextFactory.CreateDbContext();
        order.OrderNo = (db.OrderRows.Max(o => (int?)o.OrderNo) ?? 999) + 1;
        db.OrderRows.Add(ToOrderDbRecord(order));
        db.SaveChanges();
        return order;
    }

    public Order? GetOrderByOrderNo(int orderNo)
    {
        using var db = dbContextFactory.CreateDbContext();
        var order = db.OrderRows.FirstOrDefault(o => o.OrderNo == orderNo);
        return order is null ? null : ToOrderModel(order);
    }

    public Order UpdateOrder(Order order, OrderStatus? status = null, string? deliveryAddress = null,
        DeliveryMethod? deliveryMethod = null, OrderCollectionDetails? collectionDetails = null)
    {
        if (status is not null)
        {
            order.Status = status.Value;
        }

        if (deliveryAddress is not null)
        {
            order.DeliveryAddress = deliveryAddress;
        }

        if (deliveryMethod is not null)
        {
            order.DeliveryMethod = deliveryMethod.Value;
        }

        if (collectionDetails is not null)
        {
            order.OrderCollectionDetails = collectionDetails;
        }

        using var db = dbContextFactory.CreateDbContext();
        var dbOrder = db.OrderRows.FirstOrDefault(o => o.OrderNo == order.OrderNo);
        if (dbOrder is not null)
        {
            dbOrder.CustomerName = order.CustomerName;
            dbOrder.CustomerEmail = order.CustomerEmail;
            dbOrder.ItemsJson = JsonSerializer.Serialize(order.Items);
            dbOrder.DeliveryAddress = order.DeliveryAddress;
            dbOrder.Status = order.Status;
            dbOrder.StatusMessage = order.StatusMessage;
            dbOrder.PlacingTime = order.PlacingTime;
            dbOrder.Price = order.Price;
            dbOrder.DeliveryMethod = order.DeliveryMethod;
            dbOrder.CollectionDetailsJson = order.OrderCollectionDetails is null
                ? null
                : JsonSerializer.Serialize(order.OrderCollectionDetails);
            db.SaveChanges();
        }

        return order;
    }

    public IReadOnlyCollection<Order> GetOrdersByStatus(OrderStatus orderStatus)
    {
        using var db = dbContextFactory.CreateDbContext();
        return db.OrderRows
            .Where(o => o.Status == orderStatus)
            .AsEnumerable()
            .Select(ToOrderModel)
            .ToArray();
    }

    public void InsertDailyPizzaSalesEntries(IEnumerable<DailyPizzaSales> salesEntries)
    {
        using var db = dbContextFactory.CreateDbContext();
        db.DailyPizzaSalesRows.AddRange(salesEntries.Select(entry => new DailyPizzaSalesDbRecord
        {
            Id = Guid.NewGuid(),
            Date = entry.Date,
            Pizza = entry.Pizza,
            Sales = entry.Sales
        }));
        db.SaveChanges();
    }

    public IReadOnlyCollection<DailyPizzaSales> GetDailyPizzaSalesBetweenDates(DateOnly startDay, DateOnly endDay)
    {
        using var db = dbContextFactory.CreateDbContext();
        return db.DailyPizzaSalesRows
            .Where(e => e.Date >= startDay && e.Date <= endDay)
            .Select(e => new DailyPizzaSales
            {
                Date = e.Date,
                Pizza = e.Pizza,
                Sales = e.Sales
            })
            .ToArray();
    }

    public void InsertMenuItem(MenuItem item)
    {
        var db = dbContextFactory.CreateDbContext();
        db.MenuRows.Add(new MenuDbRecord
        {
            Name = item.Name,
            Price = item.Price,
            Calories = item.Calories,
            Vegetarian = item.Vegetarian
        });
        db.SaveChanges();
    }

    public void SetMenuItems(IEnumerable<MenuItem> menuItems)
    {
        using var db = dbContextFactory.CreateDbContext();
        db.MenuRows.RemoveRange(db.MenuRows);

        var rowsToInsert = menuItems
            .Select(item => new MenuDbRecord
            {
                Name = item.Name,
                Price = item.Price,
                Calories = item.Calories,
                Vegetarian = item.Vegetarian
            })
            .ToList();
        db.MenuRows.AddRange(rowsToInsert);
        db.SaveChanges();
    }

    public IReadOnlyCollection<MenuItem> GetMenuItems()
    {
        using var db = dbContextFactory.CreateDbContext();
        return db.MenuRows
            .Select(menuDbRecord => new MenuItem
            {
                Name = menuDbRecord.Name,
                Price = menuDbRecord.Price,
                Calories = menuDbRecord.Calories,
                Vegetarian = menuDbRecord.Vegetarian
            })
            .ToArray();
    }

    private static OrderDbRecord ToOrderDbRecord(Order order)
    {
        return new OrderDbRecord
        {
            OrderNo = order.OrderNo,
            CustomerName = order.CustomerName,
            CustomerEmail = order.CustomerEmail,
            ItemsJson = JsonSerializer.Serialize(order.Items),
            DeliveryAddress = order.DeliveryAddress,
            Status = order.Status,
            StatusMessage = order.StatusMessage,
            PlacingTime = order.PlacingTime,
            Price = order.Price,
            DeliveryMethod = order.DeliveryMethod,
            CollectionDetailsJson = order.OrderCollectionDetails is null
                ? null
                : JsonSerializer.Serialize(order.OrderCollectionDetails)
        };
    }

    private static Order ToOrderModel(OrderDbRecord row)
    {
        return new Order
        {
            OrderNo = row.OrderNo,
            CustomerName = row.CustomerName,
            CustomerEmail = row.CustomerEmail,
            Items = JsonSerializer.Deserialize<List<PizzaItem>>(row.ItemsJson) ?? [],
            DeliveryAddress = row.DeliveryAddress,
            Status = row.Status,
            StatusMessage = row.StatusMessage,
            PlacingTime = row.PlacingTime,
            Price = row.Price,
            DeliveryMethod = row.DeliveryMethod,
            OrderCollectionDetails = row.CollectionDetailsJson is null
                ? null
                : JsonSerializer.Deserialize<OrderCollectionDetails>(row.CollectionDetailsJson)
        };
    }
}

File: CucumberJVM/Shared/WIMP.App/java/com/wimp/app/data/DataRepository.java
/*
 * NOTE: This application ("WIMP - Where Is My Pizza") is provided solely to
 * demonstrate the Behavior-Driven Development scenario automation patterns
 * described in the book "Effective Behavior-Driven Development" by
 * Gaspar Nagy and Seb Rose.
 *
 * It is NOT a complete or production-ready implementation. It deliberately
 * uses shortcuts and simplifications (e.g. authentication, data storage,
 * error handling, security) that are NOT suitable for a real application.
 * Do not use this code as a basis for production software.
 */

package com.wimp.app.data;

import com.wimp.app.data.entities.ContactDetailsEmbeddable;
import com.wimp.app.data.entities.CouponEntity;
import com.wimp.app.data.entities.CustomerEntity;
import com.wimp.app.data.entities.DailyPizzaSalesEntity;
import com.wimp.app.data.entities.MenuItemEntity;
import com.wimp.app.data.entities.NotificationEntity;
import com.wimp.app.data.entities.OrderCollectionDetailsEmbeddable;
import com.wimp.app.data.entities.OrderEntity;
import com.wimp.app.data.entities.PizzaItemEmbeddable;
import com.wimp.app.data.entities.PromotionEntity;
import com.wimp.app.data.entities.SessionEntity;
import com.wimp.app.models.ContactDetails;
import com.wimp.app.models.Coupon;
import com.wimp.app.models.Customer;
import com.wimp.app.models.DailyPizzaSales;
import com.wimp.app.models.DeliveryMethod;
import com.wimp.app.models.MenuItem;
import com.wimp.app.models.Notification;
import com.wimp.app.models.Order;
import com.wimp.app.models.OrderCollectionDetails;
import com.wimp.app.models.OrderStatus;
import com.wimp.app.models.PizzaItem;
import com.wimp.app.models.Promotion;
import com.wimp.app.models.Session;
import jakarta.persistence.EntityManager;
import jakarta.persistence.PersistenceContext;
import org.springframework.stereotype.Repository;

import java.time.LocalDate;
import java.util.List;
import java.util.Locale;

@Repository
public class DataRepository {
    @PersistenceContext
    private EntityManager entityManager;

    public Customer getCustomerByName(String customerName) {
        return entityManager.createQuery(
                "select c from CustomerEntity c where lower(c.name) = :name",
                CustomerEntity.class)
            .setParameter("name", customerName.toLowerCase(Locale.ROOT))
            .getResultStream()
            .findFirst()
            .map(this::toCustomer)
            .orElse(null);
    }

    public void insertCustomer(Customer customer) {
        entityManager.persist(toEntity(customer));
    }

    public void insertSession(Session session) {
        entityManager.persist(toEntity(session));
    }

    public void deleteSession(String token) {
        SessionEntity entity = entityManager.find(SessionEntity.class, token);
        if (entity != null) {
            entityManager.remove(entity);
        }
    }

    public Session getSessionByToken(String token) {
        SessionEntity entity = entityManager.find(SessionEntity.class, token);
        return entity == null ? null : toSession(entity);
    }

    public Order insertOrder(Order order) {
        OrderEntity entity = toEntity(order);
        entityManager.persist(entity);
        entityManager.flush();
        order.setOrderNo(entity.getOrderNo());
        return order;
    }

    public Order getOrderByOrderNo(int orderNo) {
        OrderEntity entity = entityManager.find(OrderEntity.class, orderNo);
        return entity == null ? null : toOrder(entity);
    }

    public Order updateOrder(Order order, OrderStatus status, String deliveryAddress, DeliveryMethod deliveryMethod, OrderCollectionDetails collectionDetails) {
        OrderEntity entity = entityManager.find(OrderEntity.class, order.getOrderNo());
        if (entity == null) {
            return order;
        }

        if (status != null) {
            entity.setStatus(status);
            order.setStatus(status);
        }
        if (deliveryAddress != null) {
            entity.setDeliveryAddress(deliveryAddress);
            order.setDeliveryAddress(deliveryAddress);
        }
        if (deliveryMethod != null) {
            entity.setDeliveryMethod(deliveryMethod);
            order.setDeliveryMethod(deliveryMethod);
        }
        if (collectionDetails != null) {
            entity.setOrderCollectionDetails(toEmbeddable(collectionDetails));
            order.setOrderCollectionDetails(collectionDetails);
        }

        entityManager.merge(entity);
        return order;
    }

    public List<Order> getOrdersByStatus(OrderStatus orderStatus) {
        return entityManager.createQuery(
                "select o from OrderEntity o where o.status = :status order by o.placingTime",
                OrderEntity.class)
            .setParameter("status", orderStatus)
            .getResultList()
            .stream()
            .map(this::toOrder)
            .toList();
    }

    public void insertNotification(Notification notification) {
        entityManager.persist(toEntity(notification));
    }

    public List<Notification> getNotificationsByCustomerName(String customerName) {
        return entityManager.createQuery(
                "select n from NotificationEntity n where lower(n.customerName) = :name",
                NotificationEntity.class)
            .setParameter("name", customerName.toLowerCase(Locale.ROOT))
            .getResultList()
            .stream()
            .map(this::toNotification)
            .toList();
    }

    public void insertCoupon(Coupon coupon) {
        entityManager.persist(toEntity(coupon));
    }

    public List<Coupon> getCouponsByEmail(String customerEmail) {
        return entityManager.createQuery(
                "select c from CouponEntity c where lower(c.customerEmail) = :email",
                CouponEntity.class)
            .setParameter("email", customerEmail.toLowerCase(Locale.ROOT))
            .getResultList()
            .stream()
            .map(this::toCoupon)
            .toList();
    }

    public void insertPromotion(Promotion promotion) {
        entityManager.persist(toEntity(promotion));
    }

    public Promotion getPromotionByName(String name) {
        return entityManager.createQuery(
                "select p from PromotionEntity p where lower(p.name) = :name",
                PromotionEntity.class)
            .setParameter("name", name.toLowerCase(Locale.ROOT))
            .getResultStream()
            .findFirst()
            .map(this::toPromotion)
            .orElse(null);
    }

    public Promotion updatePromotion(Promotion promotion, Boolean isActive) {
        PromotionEntity entity = entityManager.find(PromotionEntity.class, promotion.getName());
        if (entity == null) {
            return promotion;
        }

        if (isActive != null) {
            entity.setActive(isActive);
            promotion.setActive(isActive);
        }

        entityManager.merge(entity);
        return promotion;
    }

    public void insertDailyPizzaSalesEntries(List<DailyPizzaSales> salesEntries) {
        for (DailyPizzaSales salesEntry : salesEntries) {
            entityManager.persist(toEntity(salesEntry));
        }
    }

    public List<DailyPizzaSales> getDailyPizzaSalesBetweenDates(LocalDate startDay, LocalDate endDay) {
        return entityManager.createQuery(
                "select d from DailyPizzaSalesEntity d where d.date between :startDay and :endDay",
                DailyPizzaSalesEntity.class)
            .setParameter("startDay", startDay)
            .setParameter("endDay", endDay)
            .getResultList()
            .stream()
            .map(this::toDailyPizzaSales)
            .toList();
    }

    public void insertMenuItem(MenuItem menuItem) {
        entityManager.persist(toEntity(menuItem));
    }

    public void setMenuItems(List<MenuItem> menuItems) {
        entityManager.createQuery("delete from MenuItemEntity").executeUpdate();
        for (MenuItem menuItem : menuItems) {
            entityManager.persist(toEntity(menuItem));
        }
    }

    public List<MenuItem> getMenuItems() {
        return entityManager.createQuery("select m from MenuItemEntity m order by m.name", MenuItemEntity.class)
            .getResultList()
            .stream()
            .map(this::toMenuItem)
            .toList();
    }

    private CustomerEntity toEntity(Customer customer) {
        CustomerEntity entity = new CustomerEntity();
        entity.setName(customer.name());
        entity.setEmail(customer.email());
        return entity;
    }

    private Customer toCustomer(CustomerEntity entity) {
        return new Customer(entity.getName(), entity.getEmail());
    }

    private SessionEntity toEntity(Session session) {
        SessionEntity entity = new SessionEntity();
        entity.setToken(session.token());
        entity.setCustomerName(session.customerName());
        return entity;
    }

    private Session toSession(SessionEntity entity) {
        return new Session(entity.getToken(), entity.getCustomerName());
    }

    private OrderEntity toEntity(Order order) {
        OrderEntity entity = new OrderEntity();
        entity.setOrderNo(order.getOrderNo() == 0 ? null : order.getOrderNo());
        entity.setCustomerName(order.getCustomerName());
        entity.setCustomerEmail(order.getCustomerEmail());
        entity.setItems(order.getItems().stream().map(this::toEmbeddable).toList());
        entity.setDeliveryAddress(order.getDeliveryAddress());
        entity.setStatus(order.getStatus());
        entity.setStatusMessage(order.getStatusMessage());
        entity.setPlacingTime(order.getPlacingTime());
        entity.setExpectedDeliveryTime(order.getExpectedDeliveryTime());
        entity.setPrice(order.getPrice());
        entity.setDeliveryMethod(order.getDeliveryMethod());
        entity.setOrderCollectionDetails(order.getOrderCollectionDetails() == null ? null : toEmbeddable(order.getOrderCollectionDetails()));
        entity.setContactDetails(order.getContactDetails() == null ? null : toEmbeddable(order.getContactDetails()));
        return entity;
    }

    private Order toOrder(OrderEntity entity) {
        Order order = new Order();
        order.setOrderNo(entity.getOrderNo());
        order.setCustomerName(entity.getCustomerName());
        order.setCustomerEmail(entity.getCustomerEmail());
        order.setItems(entity.getItems().stream().map(this::toDomain).toList());
        order.setDeliveryAddress(entity.getDeliveryAddress());
        order.setStatus(entity.getStatus());
        order.setStatusMessage(entity.getStatusMessage());
        order.setPlacingTime(entity.getPlacingTime());
        order.setExpectedDeliveryTime(entity.getExpectedDeliveryTime());
        order.setPrice(entity.getPrice());
        order.setDeliveryMethod(entity.getDeliveryMethod());
        order.setOrderCollectionDetails(entity.getOrderCollectionDetails() == null ? null : toDomain(entity.getOrderCollectionDetails()));
        order.setContactDetails(entity.getContactDetails() == null ? null : toDomain(entity.getContactDetails()));
        return order;
    }

    private NotificationEntity toEntity(Notification notification) {
        NotificationEntity entity = new NotificationEntity();
        entity.setCustomerName(notification.customerName());
        entity.setMessage(notification.message());
        entity.setSentAt(notification.sentAt());
        return entity;
    }

    private Notification toNotification(NotificationEntity entity) {
        return new Notification(entity.getCustomerName(), entity.getMessage(), entity.getSentAt());
    }

    private CouponEntity toEntity(Coupon coupon) {
        CouponEntity entity = new CouponEntity();
        entity.setCode(coupon.getCode());
        entity.setCustomerEmail(coupon.getCustomerEmail());
        entity.setSentAt(coupon.getSentAt());
        return entity;
    }

    private Coupon toCoupon(CouponEntity entity) {
        Coupon coupon = new Coupon();
        coupon.setCode(entity.getCode());
        coupon.setCustomerEmail(entity.getCustomerEmail());
        coupon.setSentAt(entity.getSentAt());
        return coupon;
    }

    private PromotionEntity toEntity(Promotion promotion) {
        PromotionEntity entity = new PromotionEntity();
        entity.setName(promotion.getName());
        entity.setActive(promotion.isActive());
        return entity;
    }

    private Promotion toPromotion(PromotionEntity entity) {
        Promotion promotion = new Promotion();
        promotion.setName(entity.getName());
        promotion.setActive(entity.isActive());
        return promotion;
    }

    private DailyPizzaSalesEntity toEntity(DailyPizzaSales sales) {
        DailyPizzaSalesEntity entity = new DailyPizzaSalesEntity();
        entity.setDate(sales.getDate());
        entity.setPizza(sales.getPizza());
        entity.setSales(sales.getSales());
        return entity;
    }

    private DailyPizzaSales toDailyPizzaSales(DailyPizzaSalesEntity entity) {
        DailyPizzaSales sales = new DailyPizzaSales();
        sales.setDate(entity.getDate());
        sales.setPizza(entity.getPizza());
        sales.setSales(entity.getSales());
        return sales;
    }

    private MenuItemEntity toEntity(MenuItem menuItem) {
        MenuItemEntity entity = new MenuItemEntity();
        entity.setName(menuItem.getName());
        entity.setPrice(menuItem.getPrice());
        entity.setCalories(menuItem.getCalories());
        entity.setVegetarian(menuItem.isVegetarian());
        return entity;
    }

    private MenuItem toMenuItem(MenuItemEntity entity) {
        MenuItem menuItem = new MenuItem();
        menuItem.setName(entity.getName());
        menuItem.setPrice(entity.getPrice());
        menuItem.setCalories(entity.getCalories());
        menuItem.setVegetarian(entity.isVegetarian());
        return menuItem;
    }

    private PizzaItemEmbeddable toEmbeddable(PizzaItem item) {
        PizzaItemEmbeddable embeddable = new PizzaItemEmbeddable();
        embeddable.setName(item.name());
        embeddable.setSize(item.size());
        embeddable.setStyle(item.style());
        return embeddable;
    }

    private PizzaItem toDomain(PizzaItemEmbeddable item) {
        return new PizzaItem(
            item.getName(),
            item.getSize(),
            item.getStyle());
    }

    private OrderCollectionDetailsEmbeddable toEmbeddable(OrderCollectionDetails details) {
        OrderCollectionDetailsEmbeddable embeddable = new OrderCollectionDetailsEmbeddable();
        embeddable.setOrderNo(details.getOrderNo());
        embeddable.setBoxesToBeCollected(details.getBoxesToBeCollected());
        embeddable.setContactDetailsConfirmationRequested(details.isContactDetailsConfirmationRequested());
        return embeddable;
    }

    private OrderCollectionDetails toDomain(OrderCollectionDetailsEmbeddable details) {
        OrderCollectionDetails collectionDetails = new OrderCollectionDetails();
        collectionDetails.setOrderNo(details.getOrderNo());
        collectionDetails.setBoxesToBeCollected(details.getBoxesToBeCollected());
        collectionDetails.setContactDetailsConfirmationRequested(Boolean.TRUE.equals(details.getContactDetailsConfirmationRequested()));
        return collectionDetails;
    }

    private ContactDetailsEmbeddable toEmbeddable(ContactDetails details) {
        ContactDetailsEmbeddable embeddable = new ContactDetailsEmbeddable();
        embeddable.setName(details.getName());
        embeddable.setEmail(details.getEmail());
        embeddable.setPhone(details.getPhone());
        embeddable.setCountry(details.getCountry());
        embeddable.setState(details.getState());
        return embeddable;
    }

    private ContactDetails toDomain(ContactDetailsEmbeddable details) {
        ContactDetails contactDetails = new ContactDetails();
        contactDetails.setName(details.getName());
        contactDetails.setEmail(details.getEmail());
        contactDetails.setPhone(details.getPhone());
        contactDetails.setCountry(details.getCountry());
        contactDetails.setState(details.getState());
        return contactDetails;
    }
}
